homeassistant/helpers/entity.py

```diff
@@ -218,8 +219,134 @@ class EntityPlatformState(Enum):
+            dataclass_bases.append(getattr(base, "_dataclass", base))
+        cls._dataclass = dataclasses.make_dataclass(
+            f"{name}_dataclass", class_fields, bases=tuple(dataclass_bases), frozen=True
+        )
+
+    def __new__(
+        mcs,  # noqa: N804  ruff bug, ruff does not understand this is a metaclass
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        frozen_or_thawed: bool = False,
+        **kwargs: Any,
+    ) -> Any:
+        """Pop frozen_or_thawed and store it in the namespace."""
+        namespace["_FrozenOrThawed__frozen_or_thawed"] = frozen_or_thawed
+        return super().__new__(mcs, name, bases, namespace)
+
+    def __init__(
+        cls,
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        **kwargs: Any,
+    ) -> None:
+        """Optionally create a dataclass and store it in cls._dataclass.
+
+        A dataclass will be created if frozen_or_thawed is set, if not we assume the
+        class will be a real dataclass, i.e. it's decorated with @dataclass.
+        """
+        if not namespace["_FrozenOrThawed__frozen_or_thawed"]:
+            parent = cls.__mro__[1]
+            # This class is a real dataclass, optionally inject the parent's annotations
+            if dataclasses.is_dataclass(parent) or not hasattr(parent, "_dataclass"):
+                # Rely on dataclass inheritance
+                return
+            # Parent is not a dataclass, inject its annotations
+            cls.__annotations__ = (
+                parent._dataclass.__annotations__ | cls.__annotations__
+            )
+            return
+
+        # First try without setting the kw_only flag, and if that fails, try setting it
+        try:
+            cls._make_dataclass(name, bases, False)
+        except TypeError:
+            cls._make_dataclass(name, bases, True)
+
+        def __delattr__(self: EntityDescription, name: str) -> None:
+            """Delete an attribute.
+
+            If self is a real dataclass, this is called if the dataclass is not frozen.
+            If self is not a real dataclass, forward to cls._dataclass.__delattr.
+            """
+            if dataclasses.is_dataclass(self):
+                return object.__delattr__(self, name)
+            return self._dataclass.__delattr__(self, name)  # type: ignore[unreachable]
+
+        def __setattr__(self: EntityDescription, name: str, value: Any) -> None:
+            """Set an attribute.
+
+            If self is a real dataclass, this is called if the dataclass is not frozen.
+            If self is not a real dataclass, forward to cls._dataclass.__setattr__.
+            """
+            if dataclasses.is_dataclass(self):
+                return object.__setattr__(self, name, value)
+            return self._dataclass.__setattr__(self, name, value)  # type: ignore[unreachable]
+
+        # Set generated dunder methods from the dataclass
```

Comment:
We can't expose `dataclass._FIELDS` because it's then not possible to make a child class either frozen or mutable.

This means the `dataclasses` helpers which accept either dataclass instances or dataclass classes, `dataclass.fields` and `dataclass.is_dataclass`, will only work on dataclass instances.